File: go-backend/agents/NotificationAgent.go
package agents

import (
    "fmt"
    "time"
    "github.com/neuroedge/go-backend/core"
)

type NotificationAgent struct{}

func NewNotificationAgent() *NotificationAgent { return &NotificationAgent{} }

func (a *NotificationAgent) Name() string { return "NotificationAgent" }

func (a *NotificationAgent) Init() error {
    eb := core.GetEventBus()
    eb.Subscribe("db:update", func(topic string, data interface{}) { fmt.Println("NotificationAgent received db:update", data) })
    return nil
}

func (a *NotificationAgent) Run(payload map[string]interface{}) (map[string]interface{}, error) {
    res := map[string]interface{}{"agent": "NotificationAgent", "received": payload, "ts": time.Now().UTC().String()}
    return res, nil
}

func (a *NotificationAgent) Metrics() map[string]interface{} {
    return map[string]interface{}{"name":"NotificationAgent", "uptime": 1}
}

func init() { core.RegisterAgent(NewNotificationAgent()) }

File: python-backend/backend-python/agents/NotificationAgent.py
# backend-python/agents/NotificationAgent.py

from ..core.dbManager import db
from ..core.eventBus import eventBus
from ..utils.logger import logger
import time

class NotificationAgent:
    name = "NotificationAgent"

    def __init__(self):
        # Subscribe to DB events
        eventBus.subscribe("db:update", self.handle_db_update)
        eventBus.subscribe("db:delete", self.handle_db_delete)

    async def send_notification(self, user_id: str, message: str, metadata: dict = None) -> dict:
        """
        Send a notification to a user and save it in the DB.
        """
        notification = {
            "user_id": user_id,
            "message": message,
            "metadata": metadata or {},
            "timestamp": time.time(),
            "status": "sent"
        }

        record_id = f"notif_{int(time.time()*1000)}"
        await db.set("notifications", record_id, notification, target="edge")
        eventBus.publish("db:update", {
            "collection": "notifications",
            "key": record_id,
            "value": notification,
            "source": self.name
        })

        logger.log(f"[NotificationAgent] Notification sent: notifications:{record_id}")
        return {"id": record_id, "notification": notification}

    async def handle_db_update(self, event: dict):
        logger.log(f"[NotificationAgent] DB update received: {event.get('collection')}:{event.get('key')}")

    async def handle_db_delete(self, event: dict):
        logger.log(f"[NotificationAgent] DB delete received: {event.get('collection')}:{event.get('key')}")

    async def recover(self, error: Exception):
        logger.error(f"[NotificationAgent] Recovering from error: {error}")
